client/asset/btc/btc.go

```diff
@@ -894,132 +918,248 @@ func (btc *ExchangeWallet) LocktimeExpired(contract dex.Bytes) (bool, error) {
+	// If a client wishes to repeat find redemption attempts for any
+	// of the deleted requests, client may re-call btc.FindRedemption.
+	errorAll := func(err error) {
+		for _, req := range processingQueue {
+			req.resultChan <- asset.FindRedemptionResult{
+				ContractCoinID: toCoinID(req.contractTxHash, req.contractTxOut),
+				Err:            err,
+			}
+		}
+		processingQueue = make(map[string]findRedemptionReq)
+	}
+
+	// findRedeemsInTx checks if any input of the provided tx spends any
+	// contract from the processing queue. If any such spending is found,
+	// the spent contract is deleted from the processing queue, the spending
+	// input's sig is parsed to retreive the secret and the secret is sent
+	// to the caller via the channel for registered find redemption results.
+	// If an error is encountered while trying to extract the secret, the
+	// error is communicated to the requester via the registered result chan.
+	findRedeemsInTx := func(tx *btcjson.TxRawResult) {
+		for i, input := range tx.Vin {
+			for k, req := range processingQueue {
+				if input.Txid != req.contractTxHash.String() || input.Vout != req.contractTxOut {
+					// input doesn't spend this contract, check next contract.
+					continue
+				}
+				// Found contract spender, delete from map and parse out secret.
+				delete(processingQueue, k)
+				var sigScript, secret []byte
+				redeemTxHash, err := chainhash.NewHashFromStr(tx.Txid)
+				if err == nil {
+					sigScript, err = hex.DecodeString(input.ScriptSig.Hex)
+				}
+				if err == nil {
+					secret, err = dexbtc.FindKeyPush(sigScript, req.contractHash, btc.chainParams)
+				}
+				if err != nil {
+					req.resultChan <- asset.FindRedemptionResult{
+						ContractCoinID: toCoinID(req.contractTxHash, req.contractTxOut),
+						Err:            fmt.Errorf("error extracting key from %s:%d, %v", tx.Txid, i, err),
 					}
-					key, err := dexbtc.FindKeyPush(sigScript, contractHash, btc.chainParams)
-					if err != nil {
-						return nil, fmt.Errorf("found spending input at %s:%d, but error extracting key from '%s': %v",
-							txHash, i, vin.ScriptSig.Hex, err)
+				} else {
+					req.resultChan <- asset.FindRedemptionResult{
+						ContractCoinID:   toCoinID(req.contractTxHash, req.contractTxOut),
+						RedemptionCoinID: toCoinID(redeemTxHash, uint32(i)),
+						Secret:           secret,
+						Err:              err,
 					}
-					return key, nil
 				}
+				break // skip checking other contracts for this input and check next input
 			}
 		}
-		return nil, fmt.Errorf("no key found")
-	}
-	// Since it's not a mempool transaction, start scanning blocks at the height
-	// of the swap.
-	blockHash, err := chainhash.NewHashFromStr(tx.BlockHash)
-	if err != nil {
-		return nil, fmt.Errorf("error decoding block hash: %v", err)
 	}
-	contractBlock := *blockHash
-	var contractHash []byte
-	blockHeight := tx.BlockIndex
+
+	var lastSearchedBlockHeight int64 = -1
+	var nextBlockToSearch *block
+
 	for {
```

Comment:
This loop presently just spins, not blocking or waiting for any signals aside from context cancellation.